### ISSUE TITLE
Rule proposal: no-commented-tests

### ISSUE THREAD

**Issue by Lakitna** [lo1tuma/eslint-plugin-mocha#202]
There are roughly two ways of skipping a test. 

- The proper way by using `.skip` where the test will still appear in reports.
- The bad way by commenting out the test code where the tests will simply disappear everywhere.

I previously encountered a situation in my team where a dev commented out tests, that were subsequently forgotten. Months later we found out that the code that failed these tests caused issues in production.

I would like to disallow commenting out tests to prevent people from skipping the wrong way. Bonus points if the error message tells the user to use `.skip` instead of using comments.

**Comment by lo1tuma** [comment]
This is quite hard to implement. See also [this feature request](https://github.com/eslint/eslint/issues/2735) in eslint.
The problem is that the commented code might not even be completely valid, for example:

```js
// doesn’t work right now
// it('should work', () => {
//    assert(foo());
// });
```

The first line is no valid JavaScript code, so parsing of the whole comment block would fail and we can’t determine if it actually is test code or not.

**Comment by Lakitna** [comment]
Hmm that is a shame.

After reading the link you added I see where it goes wrong. But it also might work because we can target some very specific keywords like ``/it\(['"`]/`` or ``/describe\(['"`]/``. We don't really care about incomplete code or language edge cases here.

Though maybe my optimism comes from inexperience with eslint.

**Comment by lo1tuma** [comment]
IMHO it is really hard trying to understand the intent of a comment. Think about a in-source-code documentation about how to test, e.g.

```js
// This is how e2e UI tests should look like
// it('should work', async () => {
//   const browser = await startBrowser();
//   try {
//     await browser.authenticate(defaultCredentials);
//
//     // Your test code
//   } finally {
//      await browser.close();
// }
```

**Comment by Lakitna** [comment]
Good point. And adding `// eslint-disable-next-line mocha/no-commented-tests` to that would only make things worse.

Can you make a distinction between comment blocks `/* */` and single-line comments `//`? That way you can catch the mistakes where you quickly comment out a test while maintaining an escape hatch for documentation.

**Comment by lo1tuma** [comment]
> Can you make a distinction between comment blocks `/* */` and single-line comments `//`?

Yes, that’s possible. However I don’t think it is an indicator whether it is a quickly commented out test or not. Since most IDEs can also insert those line comments very quickly even for large code blocks.

**Comment by Lakitna** [comment]
I'm out of ideas.

It's a real shame there is no good way to do this. It could catch some potentially expensive mistakes.

**Comment by Lakitna** [comment]
I had a pretty decent idea. How about failing on commented out `it` functions like mentioned before, but only when it's inside a `describe(`.

This way you can still effectively document things outside the suite.

**Comment by lo1tuma** [comment]
Sorry for the late response. I think we could ship such a rule as long as it is not part of the recommended config and its potential weaknesses are documented.

**Comment by lo1tuma** [comment]
Closing for now, since there haven’t been a good proposal yet.
